DirectProgramming/DPC++FPGA/Tutorials/Features/prototype/CSR_and_streaming_interfaces/README.md

```diff
@@ -0,0 +1,300 @@
+
+### Declaring a CSR Agent kernel interface
+
+#### Example Functor
+
+```c++
+struct MyIP {
+  MyIP() {}
+  register_map_interface void operator()() const {
+    ...
+  }
+};
+```
+
+#### Example Lambda
+```c++
+q.single_task([=] register_map_interface {
+  ...
+})
+```
+
+### Declaring a streaming kernel interface
+
+#### Example Functor
+
+```c++
+struct MyIP {
+  MyIP() {}
+  streaming_interface void operator()() const {
+    ...
+  }
+};
+```
+
+#### Example Lambda
+```c++
+q.single_task([=] streaming_interface {
+  ...
+})
+```
+
+### Declaring a CSR Agent kernel argument interface
+
+#### Example Functor
+
+```c++
+struct MyIP {
+  register_map int arg1;
+  MyIP(int arg1_) : arg1(arg1_) {}
+  register_map_interface void operator()() const {
+    ...
+  }
+};
+```
+
+*__Note__:* CSR agent kernel arguments are not currently supported in kernels with streaming control.
+
+### Declaring a streaming kernel argument interface
+
+```c++
+struct MyIP {
+  conduit int arg1;
+  MyIP(int arg1_) : arg1(arg1_) {}
+  register_map_interface void operator()() const {
+    ...
+  }
+};
+```
+
+### Default Interfaces
+If no annotation is specified for the kernel control, then a CSR agent kernel ontrol interface will be inferred by the compiler. If no annotation is specified for the a kernel argument, then that kernel argument will have the same interface as the kernel control interface. In the Lambda programming model, all kernel arguments will have the same interface as the kernel control interface.
+
+### Testing the Tutorial
+In `CSR_and_streaming_interfaces.cpp`, two kernels are decalred using the functor programming model, one with CSR Agent kernel control interface, and one with streaming kernel control interface.
+
+```c++
+struct CSRAgentControlIP {
+  register_map ValueT *input;
+  ValueT *output;
+  conduit size_t n;
+  CSRAgentControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  register_map_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+struct StreamingControlIP {
+  conduit ValueT *input;
+  conduit ValueT *output;
+  size_t n;
+  StreamingControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  streaming_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+```
+
+These two kernels are invoked in the same way in the host code, by submitting a single_task into the sycl queue.
+
+```c++
+template <typename KernelType>
+void KernelTest(sycl::queue& q, ValueT* in, ValueT* out, size_t count) {
+  q.single_task(KernelType{in, out, count}).wait();
+}
+```
+
+## Key Concepts
+* Basics of declaring kernel control interfaces and kernel argument interfaces
+
+## License
```

Comment:
Please move this License section to the very end of the README

Reply:
moved